src/Engine/Spells/CastSpellInfo.cpp

```diff
@@ -222,1501 +230,1281 @@ void CastSpellInfoHelpers::CastSpell() {
+                    pSpellSprite.field_60_distance_related_prolly_lod = target_direction.uDistance;
+                    pSpellSprite.uFacing = (short)target_direction.uYawAngle;
+                    if (pParty->bTurnBasedModeOn) {
+                        pSpellSprite.uAttributes |= SPRITE_HALT_TURN_BASED;
+                    }
+                    spell_speed = pObjectList->pObjects[pSpellSprite.uObjectDescID].uSpeed;
+                    if (pSpellSprite.Create(pParty->sRotationZ, pParty->sRotationY + 10, spell_speed, pCastSpell->uPlayerID + 1) != -1 && pParty->bTurnBasedModeOn) {
+                        ++pTurnEngine->pending_actions;
+                    }
+                    break;
                 }
-                spell_sound_flag = true;
-                break;
-            }
 
-            case SPELL_EARTH_MASS_DISTORTION:
-            {
-                monster_id = PID_ID(spell_targeted_at);
-                if (!pPlayer->CanCastSpell(uRequiredMana))
-                    break;
-                if (pActors[monster_id].DoesDmgTypeDoDamage(DMGT_EARTH)) {
-                    pActors[monster_id].pActorBuffs[ACTOR_BUFF_MASS_DISTORTION].Apply(GameTime(pMiscTimer->uTotalGameTimeElapsed + 128), PLAYER_SKILL_MASTERY_NONE, 0, 0, 0);
-                    v704.x = 0;
-                    v704.y = 0;
-                    v704.z = 0;
-                    pSpellSprite.containing_item.Reset();
-                    pSpellSprite.spell_id = pCastSpell->uSpellID;
-                    pSpellSprite.spell_level = spell_level;
-                    pSpellSprite.spell_skill = spell_mastery;
-                    pSpellSprite.uObjectDescID = pObjectList->ObjectIDByItemID(pSpellSprite.uType);
-                    pSpellSprite.uAttributes = 0;
-                    pSpellSprite.uSectorID = 0;
-                    pSpellSprite.uSpriteFrameID = 0;
-                    pSpellSprite.field_60_distance_related_prolly_lod = 0;
-                    pSpellSprite.uFacing = 0;
-                    pSpellSprite.spell_caster_pid = PID(OBJECT_Player, pCastSpell->uPlayerID);
-                    pSpellSprite.uSoundID = (short)pCastSpell->sound_id;
-                    pSpellSprite.vPosition.x = pActors[monster_id].vPosition.x;
-                    pSpellSprite.vPosition.y = pActors[monster_id].vPosition.y;
-                    pSpellSprite.vPosition.z = pActors[monster_id].vPosition.z;
-                    pSpellSprite.spell_target_pid = PID(OBJECT_Actor, monster_id);
-                    Actor::DamageMonsterFromParty(PID(OBJECT_Item, pSpellSprite.Create(0, 0, 0, 0)), monster_id, &v704);
+                case SPELL_AIR_IMPLOSION:
+                {
+                    monster_id = PID_ID(spell_targeted_at);
+                    if (!spell_targeted_at) {
+                        SpellFailed(pCastSpell, LSTR_SPELL_FAILED);
+                        pPlayer->CastSpellModifyMana(uRequiredMana); // decrease mana on failure
```

Comment:
You're introducing a bug here – `pPlayer->CanCastSpell(uRequiredMana)` call above has already subtracted the mana cost.

`CanCastSpell` is obviously not the best name, given what the function does, so please feel free to rename it.

Reply:
Also we'll need a test to catch this.

https://github.com/OpenEnroth/OpenEnroth/issues/414

Reply:
I splitted functionality of this function.
CanCastSpell just checks now if spell can be cast at all and SpendMana physically decreases player's mana.